// @openclaw/plugin-sdk/index.js
/**
 * Claw2Cli Plugin SDK Shim
 *
 * This module impersonates @openclaw/plugin-sdk so that OpenClaw plugins
 * (e.g., @tencent-weixin/openclaw-weixin) can load without the real OpenClaw runtime.
 *
 * Communication with the Go daemon is via stdin/stdout NDJSON.
 */

const { createInterface } = require("readline");
const path = require("path");
const fs = require("fs");

// ---------------------------------------------------------------------------
// NDJSON bridge to Go daemon (stdin/stdout)
// ---------------------------------------------------------------------------

const pendingRequests = new Map(); // id -> { resolve, reject }
let requestIdCounter = 0;

function sendMessage(msg) {
  msg.ts = Date.now();
  process.stdout.write(JSON.stringify(msg) + "\n");
}

function sendEvent(source, topic, payload) {
  sendMessage({ type: "event", source, topic, payload });
}

function sendLog(source, level, message) {
  sendMessage({ type: "log", source, level, message });
}

function sendCommand(source, action, payload) {
  const id = `req-${++requestIdCounter}`;
  sendMessage({ type: "command", source, action, payload, id });
  return new Promise((resolve, reject) => {
    pendingRequests.set(id, { resolve, reject });
    // Timeout after 5 minutes
    setTimeout(() => {
      if (pendingRequests.has(id)) {
        pendingRequests.delete(id);
        reject(new Error(`Request ${id} timed out`));
      }
    }, 300000);
  });
}

// Read responses from Go daemon via stdin
const rl = createInterface({ input: process.stdin, terminal: false });
rl.on("line", (line) => {
  try {
    const msg = JSON.parse(line);
    if (msg.type === "response" && msg.id && pendingRequests.has(msg.id)) {
      const { resolve } = pendingRequests.get(msg.id);
      pendingRequests.delete(msg.id);
      resolve(msg.payload);
    } else if (msg.type === "command") {
      handleInboundCommand(msg);
    }
  } catch (e) {
    // Ignore malformed input
  }
});

// Handle commands from Go daemon (e.g., "send a reply to WeChat")
let registeredChannel = null;

function handleInboundCommand(msg) {
  if (!registeredChannel) return;
  const { action, payload } = msg;

  if (action === "send_text" && registeredChannel.outbound) {
    registeredChannel.outbound
      .sendText({
        cfg: globalConfig,
        to: payload.to,
        text: payload.text,
        accountId: payload.accountId,
      })
      .then((result) => {
        sendMessage({
          type: "response",
          source: pluginSource,
          id: msg.id,
          payload: result,
        });
      })
      .catch((err) => {
        sendMessage({
          type: "error",
          source: pluginSource,
          code: "SEND_FAILED",
          message: err.message,
          id: msg.id,
        });
      });
  }
}

// ---------------------------------------------------------------------------
// Shim state
// ---------------------------------------------------------------------------

const pluginSource = process.env.C2C_PLUGIN_NAME || "unknown";
const storageDir = process.env.C2C_STORAGE_DIR || path.join(process.env.HOME, ".c2c", "storage", pluginSource);
let globalConfig = {};
let globalRuntime = null;

// ---------------------------------------------------------------------------
// Fake PluginRuntime
// ---------------------------------------------------------------------------

function buildChannelRuntime() {
  return {
    routing: {
      resolveAgentRoute({ cfg, channel, accountId, peer }) {
        // Single-agent mode: always route to default
        return {
          agentId: "c2c-default",
          sessionKey: `${accountId}:${peer.id}`,
          mainSessionKey: `${accountId}:${peer.id}`,
        };
      },
    },
    session: {
      resolveStorePath(storeConfig, { agentId }) {
        const dir = path.join(storageDir, "sessions", agentId || "default");
        fs.mkdirSync(dir, { recursive: true });
        return dir;
      },
      recordInboundSession({ storePath, sessionKey, ctx, updateLastRoute, onRecordError }) {
        // Minimal: write session record to file
        try {
          const sessionFile = path.join(storePath, "last-session.json");
          fs.writeFileSync(sessionFile, JSON.stringify({ sessionKey, timestamp: Date.now() }));
        } catch (e) {
          if (onRecordError) onRecordError(e);
        }
      },
    },
    reply: {
      finalizeInboundContext(ctx) {
        return ctx; // Pass through
      },
      resolveHumanDelayConfig(cfg, agentId) {
        return { delay: 0, jitter: 0 };
      },
      createReplyDispatcherWithTyping({ humanDelay, typingCallbacks, deliver, onError }) {
        // The dispatcher collects reply chunks and calls deliver() for each
        const dispatcher = {
          dispatch(payload) {
            try {
              deliver(payload);
            } catch (e) {
              if (onError) onError(e);
            }
          },
        };
        const replyOptions = {};
        const markDispatchIdle = () => {};
        return { dispatcher, replyOptions, markDispatchIdle };
      },
      withReplyDispatcher({ dispatcher, run }) {
        return run();
      },
      async dispatchReplyFromConfig({ ctx, cfg, dispatcher, replyOptions }) {
        // THIS IS THE KEY SHIM:
        // Instead of calling an LLM, forward the message to Go daemon and wait for a reply.
        sendEvent(pluginSource, "message.received", {
          from: ctx.From,
          to: ctx.To,
          body: ctx.Body,
          accountId: ctx.AccountId,
          messageId: ctx.MessageSid,
          chatType: ctx.ChatType,
          timestamp: ctx.Timestamp,
          mediaUrl: ctx.MediaUrl,
          mediaPath: ctx.MediaPath,
          mediaType: ctx.MediaType,
          contextToken: ctx.context_token,
        });

        // Wait for reply from Go daemon
        try {
          const reply = await sendCommand(pluginSource, "get_reply", {
            from: ctx.From,
            to: ctx.To,
            body: ctx.Body,
            accountId: ctx.AccountId,
            messageId: ctx.MessageSid,
          });

          if (reply && reply.text) {
            dispatcher.dispatch({ text: reply.text });
          }
        } catch (e) {
          sendLog(pluginSource, "error", `Failed to get reply: ${e.message}`);
        }
      },
    },
    media: {
      saveMediaBuffer({ buffer, fileName, mimeType }) {
        const mediaDir = path.join(storageDir, "media");
        fs.mkdirSync(mediaDir, { recursive: true });
        const filePath = path.join(mediaDir, fileName || `media-${Date.now()}`);
        fs.writeFileSync(filePath, buffer);
        return { filePath, url: `file://${filePath}` };
      },
    },
    commands: {},
  };
}

function buildPluginRuntime() {
  const channelRuntime = buildChannelRuntime();
  return {
    channel: channelRuntime,
    // Plugins may access other runtime properties
    config: globalConfig,
    log: (msg) => sendLog(pluginSource, "info", msg),
    error: (msg) => sendLog(pluginSource, "error", msg),
  };
}

// ---------------------------------------------------------------------------
// Exported SDK functions (matching @openclaw/plugin-sdk API)
// ---------------------------------------------------------------------------

function buildChannelConfigSchema(zodSchema) {
  return zodSchema;
}

function normalizeAccountId(raw) {
  return String(raw).replace(/[^a-zA-Z0-9_-]/g, "_").toLowerCase();
}

function withFileLock(lockPath, fn) {
  // Simplified: no actual file locking in shim
  return fn();
}

function createTypingCallbacks({ accountId, to, baseUrl, token, typingTicket, log }) {
  // Return typing start/stop callbacks
  return {
    startTyping: () => {},
    stopTyping: () => {},
    onTypingError: () => {},
  };
}

function resolveSenderCommandAuthorizationWithRuntime(runtime, ctx) {
  // Always authorized in c2c mode
  return { authorized: true };
}

function resolveDirectDmAuthorizationOutcome(ctx, config) {
  return { allowed: true };
}

function resolvePreferredOpenClawTmpDir() {
  const dir = path.join(storageDir, "tmp");
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function stripMarkdown(text) {
  if (!text) return "";
  return text
    .replace(/\*\*(.*?)\*\*/g, "$1")
    .replace(/\*(.*?)\*/g, "$1")
    .replace(/`(.*?)`/g, "$1")
    .replace(/#{1,6}\s/g, "")
    .replace(/\[([^\]]+)\]\([^)]+\)/g, "$1");
}

// ---------------------------------------------------------------------------
// Additional SDK functions (used by Lark/Feishu plugin and others)
// ---------------------------------------------------------------------------

const DEFAULT_ACCOUNT_ID = "default";
const DEFAULT_GROUP_HISTORY_LIMIT = 50;
const PAIRING_APPROVED_MESSAGE = "Pairing approved.";
const SILENT_REPLY_TOKEN = "__silent__";

function emptyPluginConfigSchema() {
  return {};
}

function isNormalizedSenderAllowed(senderId, allowList) {
  if (!allowList || allowList.length === 0) return true;
  return allowList.includes(senderId);
}

function addWildcardAllowFrom() {
  return ["*"];
}

function resolveSenderCommandAuthorization(ctx) {
  return { authorized: true };
}

function resolveThreadSessionKeys(ctx) {
  return { sessionKey: ctx.SessionKey || `${ctx.AccountId}:${ctx.From}`, threadKey: null };
}

function createReplyPrefixContext() {
  return { prefix: "", suffix: "" };
}

function buildPendingHistoryContextFromMap() {
  return {};
}

function clearHistoryEntriesIfEnabled() {}
function recordPendingHistoryEntryIfEnabled() {}
function logTypingFailure(err) {}

function formatDocsLink(path, label) {
  return label || path;
}

function buildRandomTempFilePath(ext) {
  const dir = resolvePreferredOpenClawTmpDir();
  return path.join(dir, `tmp-${Date.now()}-${Math.random().toString(36).slice(2)}${ext || ""}`);
}

function extractToolSend(result) {
  return result;
}

function jsonResult(data) {
  return JSON.stringify(data);
}

function readStringParam(params, key, defaultValue) {
  if (params && params[key] != null) return String(params[key]);
  return defaultValue || "";
}

function readReactionParams(params) {
  return { emoji: params?.emoji || "", messageId: params?.messageId || "" };
}

// ---------------------------------------------------------------------------
// OpenClawPluginApi shim
// ---------------------------------------------------------------------------

class PluginApiShim {
  constructor() {
    this.runtime = buildPluginRuntime();
    globalRuntime = this.runtime;
  }

  registerChannel({ plugin }) {
    registeredChannel = plugin;
    sendLog(pluginSource, "info", `Channel registered: ${plugin.id}`);
  }

  registerCli(callback, opts) {
    // CLI extensions are not needed in c2c mode
    sendLog(pluginSource, "debug", "CLI registration skipped in c2c mode");
  }
}

// ---------------------------------------------------------------------------
// Module exports
// ---------------------------------------------------------------------------

module.exports = {
  // Config
  buildChannelConfigSchema,
  emptyPluginConfigSchema,

  // Constants
  DEFAULT_ACCOUNT_ID,
  DEFAULT_GROUP_HISTORY_LIMIT,
  PAIRING_APPROVED_MESSAGE,
  SILENT_REPLY_TOKEN,

  // Utilities
  normalizeAccountId,
  withFileLock,
  createTypingCallbacks,
  stripMarkdown,
  formatDocsLink,
  buildRandomTempFilePath,
  extractToolSend,
  jsonResult,
  readStringParam,
  readReactionParams,
  createReplyPrefixContext,
  addWildcardAllowFrom,

  // Authorization
  resolveSenderCommandAuthorization,
  resolveSenderCommandAuthorizationWithRuntime,
  resolveDirectDmAuthorizationOutcome,
  isNormalizedSenderAllowed,
  resolvePreferredOpenClawTmpDir,

  // Session / History
  resolveThreadSessionKeys,
  buildPendingHistoryContextFromMap,
  clearHistoryEntriesIfEnabled,
  recordPendingHistoryEntryIfEnabled,
  logTypingFailure,

  // Types (no-ops at runtime, exported for destructuring safety)
  OpenClawPluginApi: PluginApiShim,

  // Internal: used by c2c-shim.js
  _internal: {
    PluginApiShim,
    sendEvent,
    sendLog,
    sendCommand,
    sendMessage,
    getRegisteredChannel: () => registeredChannel,
    getGlobalRuntime: () => globalRuntime,
    setGlobalConfig: (cfg) => { globalConfig = cfg; },
  },
};
